bayes_opt/util.py

```diff
@@ -44,38 +44,49 @@ def acq_max(ac, gp, y_max, bounds, random_state, constraint=None, n_warmup=10000
     :return: x_max, The arg max of the acquisition function.
     """
 
+    # We need to adjust the acquisition function to deal with constraints when there is some
+    if constraint is not None:
+
+        def adjusted_ac(x, gp, y_max):
```

Comment:
1. this is a somewhat minor point, but this pattern seems confusing - there is no reason for `adjusted_ac` to be inside `acq_max` is there? perhaps this is just a preference... 
2. I think we only ever intend these functions to be called from inside the code (right @till-m?) - in that case should we take this chance to append the `_` to them?

Reply:
Done

Reply:
2\. is a consequence of 1.: The function is only defined within the scope of acq max so it cannot be accessed from outside, in that sense prepending the `_` is not necessary.

Since we need to provide `minimize` with a single-argument function, we need to define a function locally (either named, like this, or anonymous, using `lambda`). This function is also only relevant for this specific optimization step -- earlier or later the information that we're working with (represented by `gp` and `y_max`) will have changed, and so will the function to optimize. In this sense, I think this function should indeed be defined here and also doesn't need an underscore. However, the two-layer usage that is introduced in this PR is a bit suboptimal IMO: First, you define `adjusted_ac` and then later you wrap that in a lambda function. Since `y_max` and `gp` don't change during the `acq_max` step, they shouldn't be used as inputs for the function and the construction of the lambda function will not be necessary.

Reply:
I see what you mean, but there is a few issues of removing the `lambda` in the `minimize`:

1) The warmup phase is a maximization problem and the second phase is a minimization problem, we could unify this behaviour and transform both problems in a minimization problem. ✅ 

2) To remove the `gp` and `y_max` from the function we would need to introduce a new lambda/function in the "else" part

```python
    if constraint is not None:

        def adjusted_ac(x, gp, y_max):
            (...)
            return values * p_constraints
    else:
        adjusted_ac = lambda x: -ac(x, gp=gp, y_max=y_max)
```

I guess one advantage is that the `minimize` part do many calls and the warmup phase does only a single call.

3) We would also need to conditionally `.reshape(1, -1` inside the `adjusted_ac`, because the `minimize` passes a 1D array if I'm not mistaken.

```python
    if constraint is not None:

        def adjusted_ac(x, gp, y_max):
            """Acquisition function adjusted to fulfill the constraint when necessary"""
            if len(x.shape) == 1:
                x = x.reshape(1, -1)

            values = ac(x, gp=gp, y_max=y_max)
            (...)
            return values * p_constraints

    else:
        adjusted_ac = lambda x: -ac(x.reshape(1, -1) if len(x.shape) == 1 else x, gp=gp, y_max=y_max)
```

What do you guys prefer?

Reply:
I removed the commit that added the `_` to the `adjusted_ac`  function.

Reply:
Something like this should work, right?
```python
    if constraint is not None:
        def adjusted_ac(x):
            """Acquisition function adjusted to fulfill the constraint when necessary"""
            values = ac(x.reshape((-1, bounds.shape[0]), gp=gp, y_max=y_max)
            p_constraints = constraint.predict(x)

            # Slower fallback for the case where any values are negative
            if np.any(values < 0):
                # TODO: This is not exactly how Gardner et al do it.
                # Their way would require the result of the acquisition function
                # to be strictly positive, which is not the case here. For a
                # positive target value, we use Gardner's version. If the target
                # is negative, we instead slightly rescale the target depending
                # on the probability estimate to fulfill the constraint.
                return np.array(
                    [
                        value * p if value > 0 else value / (0.5 + 0.5 * p)
                        for value, p in zip(values, p_constraints)
                    ]
                )

            # Faster, vectorized version of Gardner et al's method
            return values * p_constraints

    else:
        adjusted_ac = lambda x: ac(x.reshape((-1, bounds.shape[0]), gp=gp, y_max=y_max)
```

Could also think of assigning `bounds.shape[0]` to a meaningfully named variable since it's reused a few times in this section.

Reply:
Did what your suggested.

I couldn't find a good name for `bounds.shape[0]` do you have a suggestion?